src/libraries/System.Private.CoreLib/src/System/Random.cs

```diff
@@ -1,197 +1,599 @@
-        private static Random? t_threadRandom;
+        /// <summary>The underlying generator implementation.</summary>
+        /// <remarks>
+        /// This is separated out so that different generators can be used based on how this Random instance is constructed.
+        /// If it's built from a seed, then we may need to ensure backwards compatibility for folks expecting consistent sequences
+        /// based on that seed.  If the instance is actually derived from Random, then we need to ensure the derived type's
+        /// overrides are called anywhere they were being called previously.  But if the instance is the base type and is constructed
+        /// with the default constructor, we have a lot of flexibility as to how to optimize the performance and quality of the generator.
+        /// </remarks>
+        private readonly ImplBase _impl;
+
+        /// <summary>Initializes a new instance of the <see cref="Random"/> class using a default seed value.</summary>
+        public Random() =>
+            // With no seed specified, if this is the base type, we can implement this however we like.
+            // If it's a derived type, for compat we respect the previous implementation.
+            _impl = GetType() == typeof(Random) ?
+                new Xoshiro256StarStarImpl() :
+                new LegacyImpl(this);
+
+        /// <summary>Initializes a new instance of the Random class, using the specified seed value.</summary>
+        /// <param name="Seed">
+        /// A number used to calculate a starting value for the pseudo-random number sequence. If a negative number
+        /// is specified, the absolute value of the number is used.
+        /// </param>
+        public Random(int Seed) =>
+            // With a custom seed, for compat we respect the previous implementation.
+            _impl = new LegacyImpl(this, Seed);
+
+        /// <summary>Returns a non-negative random integer.</summary>
+        /// <returns>A 32-bit signed integer that is greater than or equal to 0 and less than <see cref="int.MaxValue"/>.</returns>
+        public virtual int Next() => _impl.Next();
+
+        /// <summary>Returns a non-negative random integer that is less than the specified maximum.</summary>
+        /// <param name="maxValue">The exclusive upper bound of the random number to be generated. <paramref name="maxValue"/> must be greater than or equal to 0.</param>
+        /// <returns>
+        /// A 32-bit signed integer that is greater than or equal to 0, and less than <paramref name="maxValue"/>; that is, the range of return values ordinarily
+        /// includes 0 but not <paramref name="maxValue"/>. However, if <paramref name="maxValue"/> equals 0, <paramref name="maxValue"/> is returned.
+        /// </returns>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="maxValue"/> is less than 0.</exception>
+        public virtual int Next(int maxValue)
+        {
+            if (maxValue <= 1)
```

Comment:
Can we write this as
```c#
if (maxValue > 1)
{
    return _impl.Next(maxValue);
}

// the rest
```
I think of having a fast-path and a cold-path for the <= 1 case (which should be unlikely).
(E.g. for doing monte-carlo based processes, and if BCL offers a proper random algorithm I'd like to use that more than some external algorithms).

Reply:
Based on the principle that the CPU will guess the branch is taken?

Reply:
Yeah, and in the hope that JIT gets better with devirtualization (through PGO, etc) so that we can get an inline-able fast path.
The case maxValue <= 1 is very unlikely, so this should be really cold.